packages/ui/react-ui-table/src/components/ColumnMenu.tsx

```diff
@@ -25,224 +25,157 @@ export type ColumnMenuProps<TData extends RowData, TValue> = {
 
 export const ColumnMenu = <TData extends RowData, TValue>({ column, ...props }: ColumnMenuProps<TData, TValue>) => {
   const title = column.label?.length ? column.label : column.id;
-
   const header = props.context.header;
 
-  const canSort = header.column.getCanSort();
-  const sortDirection = header.column.getIsSorted();
-  const toggleSort = header.column.getToggleSortingHandler();
+  const { canSort, sortDirection, onSelectSort, onToggleSort, onClearSort } = useColumnSorting(header.column);
+
+  const columnSettingsAnchorRef = useRef<any>(null);
+  const [isMounted, setIsMounted] = useState(false);
+
+  useEffect(() => {
+    setIsMounted(true);
+    return () => setIsMounted(false);
+  }, []);
+
+  const [isColumnSettingsOpen, setIsColumnSettingsOpen] = useState(false);
+
+  const shouldPreventFocusOnClose = useRef(false);
+
+  const onOpenColumnSettings = useCallback(() => {
+    shouldPreventFocusOnClose.current = true;
+    setIsColumnSettingsOpen(true);
+  }, [shouldPreventFocusOnClose, setIsColumnSettingsOpen]); // setIsColumnSettingsOpen is a dependency
+
+  const onDropdownCloseAutoFocus = useCallback(
+    (event: Event) => {
+      if (shouldPreventFocusOnClose.current) {
+        event.preventDefault();
+        shouldPreventFocusOnClose.current = false;
+      }
+    },
+    [shouldPreventFocusOnClose],
+  );
 
   return (
-    <div className='flex items-center gap-2'>
-      <div className='flex-1 min-is-0 truncate' title={title}>
-        {title}
-      </div>
-      <div className='flex gap-0'>
-        {canSort && toggleSort && <SortButton sortDirection={sortDirection} onClick={toggleSort} />}
-        <ColumnPanel {...props} column={column} />
+    <div className='flex items-center justify-center justify-between'>
+      <div className='flex items-center gap-1 truncate' title={title}>
+        <SortIndicator direction={sortDirection} onClick={onToggleSort} />
+        <div className='truncate'>{title}</div>
       </div>
+      {isMounted && (
+        <ColumnSettingsPanel
+          {...props}
+          column={column}
+          anchorNode={columnSettingsAnchorRef.current}
+          open={isColumnSettingsOpen}
+          setOpen={setIsColumnSettingsOpen}
+        />
+      )}
+
+      <DropdownMenu.Root modal={false}>
+        <DropdownMenu.Trigger ref={columnSettingsAnchorRef}>
```

Comment:
Ah, I see the issue — this needs `asChild` since it has a styled button within, and its parent should be `<Popover.Anchor asChild>`, extracted from `ColumnSettingsPanel`. Both DropdownMenu and Popover roots should be ancestors of all of this.

Reply:
Note that `asChild` uses Radix’s `Slot` to (mostly) cleanly compose refs and handlers.